lectures/zksnark.tex

```diff
@@ -0,0 +1,408 @@
+        x_2 \times x_3 &= mult \tag{2} \\
+        x_1 \times mult &= selectMult \tag{3} \\
+        (1 - x_1) \times (x_2 + x_3) &= r - selectMult \tag{4}
+    \end{align*}
+
+    For every constraint we need the coefficients vectors $a_i$, $b_i$, $c_i$, but all of them have
+    the same witness vector $w$.
+
+    The interesting thing is where to take a constant from. In our case, it's just 1 in the (4) 
+    constraint. The solution is straightforward: by placing 1 in the witness vector, we can then 
+    obtain any desired value by multiplying it by an appropriate coefficient which is $-1$ in our
+    case.
+
+    The witness vector:
+    \[ w = (1, r, x_1, x_2, x_3, mult, selectMult) \]
+    The coefficients vectors:
+    \begin{align*}
+        a_1 &= (0, 0, 1, 0, 0, 0, 0) & \quad b_1 &= (0, 0, 1, 0, 0, 0, 0) & \quad c_1 &= (0, 0, 1, 0, 0, 0, 0) \\
+        a_2 &= (0, 0, 0, 1, 0, 0, 0) & \quad b_2 &= (0, 0, 0, 0, 1, 0, 0) & \quad c_2 &= (0, 0, 0, 0, 0, 1, 0) \\
+        a_3 &= (0, 0, 1, 0, 0, 0, 0) & \quad b_3 &= (0, 0, 0, 0, 0, 1, 0) & \quad c_3 &= (0, 0, 0, 0, 0, 0, 1) \\
+        a_4 &= (1, 0, -1, 0, 0, 0, 0) & \quad b_4 &= (0, 0, 0, 1, 1, 0, 0) & \quad c_4 &= (0, 1, 0, 0, 0, 0, -1)
+    \end{align*}
+
+    Now, let's use specific values to compute an example:
+    \[ x_1 = 1, \quad x_2 = 3, \quad x_3 = 4 \]
+
+    We get:
+    \[ \text{mult} = x_2 \times x_3 = 3 \times 4 = 12 \]
+    \[ \text{selectMult} = x_1 \times \text{mult} = 1 \times 12 = 12 \]
+    \[ r = x_1 \times (x_2 \times x_3) + (1 - x_1) \times (x_2 + x_3) \]
+    \[ r = 1 \times (3 \times 4) + (1 - 1) \times (3 + 4) = 12 \]
+
+    Verifying the constraints:
+    \begin{enumerate}
+        \item \( x_1 \times x_1 = x_1 \quad (1 \times 1 = 1) \)
+        \item \( x_2 \times x_3 = \text{mult} \quad (3 \times 4 = 12) \)
+        \item \( x_1 \times \text{mult} = \text{selectMult} \quad (1 \times 12 = 12) \)
+        \item \( (1 - x_1) \times (x_2 + x_3) = r - \text{selectMult} \quad (0 \times 7 = 12 - 12) \)
+    \end{enumerate}
+
+    The witness vector with these values:
+    \[ w = (1, 12, 1, 3, 4, 12, 12) \]
+\end{example}
+
+Each constraint enforces that the product of the linear combinations defined by $a$ and $b$ must
+equal the linear combination defined by $c$. Collectively, these constraints describe the 
+computation by ensuring that every step, from inputs through intermediates to outputs, satisfies 
+the defined relationships, thus encoding the entire computational process in the form of a system
+of rank-1 quadratic equations.
+
+The last unresolved question is where the ``rank-1`` comes from.
+
+\begin{lemma}
+    Suppose there is a constraint $\<a, w\> \times \<b, w\> = \< c, w \>$ with coefficient vectors
+    $a$, $b$, $c$ and witness vector $w$. Then it can be expressed in the form:
+    \[ w^T A w + c^T w = 0 \]
+    Where $A$ is the outer product of vectors $a$, $b$ (denoted as $a \otimes b$), consequently 
```

Comment:
Describe what the outer product is (since only the inner product was defined previously).

Additionally, it could make sense to make a separate subsection for "Linear Algebruh Preliminaries", but it is up to you to decide